### ISSUE TITLE
Model Buildings.DHC.Plants.Combined.Controls.BaseClasses.Validation.StagingPlant fails verification due to a race condition

### ISSUE THREAD

**Issue by AndreaBartolini** [OpenModelica/OpenModelica#12619]
The following variable of the model `Buildings.DHC.Plants.Combined.Controls.BaseClasses.Validation.StagingPlant` fails the verification vs the Dymola reference file:
- `staChi.y1ChiHea[1]`

![image](https://github.com/OpenModelica/OpenModelica/assets/25974985/7eaa3269-3160-4468-be41-b46a5b47c0a1)

the variable is produced by the output `y[4]` of the `hol` block (selected in the figure) of the following logic diagram:
![image](https://github.com/OpenModelica/OpenModelica/assets/25974985/c8e76a80-343c-4c08-a547-aae87404c4dc)

the `hol` block  implements a ZOH applied to an array of integers, by means of the following algorithm:
```modelica
algorithm
  when Modelica.Math.BooleanVectors.anyTrue({
    u[i] <> pre(y[i]) for i in 1:nin}) and time - time_change > holdDuration then
    y := u;
    time_change := time;
  end when; 
```
In the analyzed case `holdDuration` = 900 seconds.

The race condition is between the inputs `u[1]` and `u[4]` of the `hol` block, that change their value conceptually @ time = 1900 seconds but really not exactly at the same time.

**herebelow the events trace from OpenModelica:**

check for discrete changes at **time=1900.00000149**
discrete var changed: **staChi.hol.time_change from 0 to 1900**
discrete var changed: **staChi.hol.u[1] from 0 to 1**
discrete var changed: staChi.hol.y[1] from 0 to 1

check for discrete changes at **time=1900.00000461**
discrete var changed: **staChi.hol.u[4] from 0 to 1**

check for discrete changes at **time=2800.00000167**
discrete var changed: **staChi.hol.time_change from 1900 to 2800**
discrete var changed: **staChi.hol.y[4] from 0 to 1**

@ time=1900.00000149 `staChi.hol.u[1]` changes from 0 to 1 and the `staChi.hol.time_change` is updated to 1900 seconds, 
@ time=1900.00000461 `staChi.hol.u[4]` changes from 0 to 1 but the `holdDuration` is not expired, so the `hol` block holds its outputs,
@ time=2800.00000167 the `holdDuration` expires and the difference between `staChi.hol.u[4]` and `staChi.hol.y[4]` is detected, so the output `staChi.hol.y[4]` is updated.

The result is that the change of the output `staChi.hol.y[4]` is delayed by one `holdDuration` time.

**herebelow the events trace from Dymola:**

Time event at time: **1900.000000369231**
  Variable **staChi.hol.time_change = 1900 at time 1900**
  Variable **staChi.hol.y[1] = 1 at time 1900**
  Variable **staChi.hol.y[4] = 1 at time 1900**

unfortunately the variables `staChi.hol.u[x]` are not reported in the Dymola log file but not any other event is triggered around 1900 seconds, so it seems that Dymola in some way is able to synchronize the signals of the entire logic diagram @ 1900 seconds (it is not detailed here but from the Dymola log file it is possible to see that all the blocks that generate events around 1900 seconds in the above diagram are triggered in the event iteration that starts @ time: 1900.000000369231).

As general conclusion, as stated in other similar tickets, the shift of one sample between OpenModelica and Dymola is possible due to different execution of asynchronous events.

**Comment by casella** [comment]
@AndreaBartolini could you open one ticket which contains a list of the tickets caused by this common issue, so we know exactly where they are and how many? You just need to list them like this:

- #12619

Thanks

**Comment by casella** [comment]
@mwetter, we analyzed this issue more in depth. As we understand, the root cause of the problem is that all discrete-time controllers are meant to generate _time events_, not _state events_, and unfortunately OMC is still not good enough at that, see good old #2152, so it handles those time events as state events with unknown triggering time.

The problem with that is not only that it is inefficient, because it requires iterations to find out the exact triggering times, but most importantly that events that should happen simultaneously get separated by very small time delays, as discussed in this ticket, which leads to incorrect behaviour every time you have a race condition.

Grouping together simultaneous time events requires first-year Hogwarts magic: even in the case their event times are not exactly coincident, you can always consider small epsilon time differences to group then appropriately. Grouping together simultaneous state event is more like Voldemort-level magic, including seeing into the future. Not really easy.

Bottom line: this issue is very likely to be fixed as soon as OMC handles events generated by relations:
- `time >= discrete_expression`
- `time < discrete expression`
as _time events_, not as _state events_, as suggested in [MLS Sect. 8.5](https://specification.modelica.org/maint/3.6/equations.html#events-and-synchronization). BTW, I wonder at this point whether this aspect can really be considered a "quality of implementation" issue, or rather a mandatory feature for a Modelica compiler. I think we should discuss this further in the MAP-Lang and MAP-Lib groups.

Fixing #2152 is also most likely going to resolve the other tickets collected in #12740, or at least most of them.

**Comment by mwetter** [comment]
@casella : I agree that handling these events as time events is important for reproducing the behavior that modelers expect for events that the model schedules to occur simultaneously.

**Comment by casella** [comment]
I hope it is not a too complicated fix, and I am quite confident it should be enough to fix a number of issues.
